### PR TITLE
Remove the OCaml emulator

### LINKED ISSUE #509: Remove OCaml emulator
Can we remove the OCaml emulator? I think this would have significant benefits:

* Less confusing for new users - they no longer need to choose between two very similar but different emulators.
* We no longer have to implement emulator features twice.
* Contributors don't need to learn OCaml.
* Shorter CI time.
* When we use the binary release of the Sail compiler, this will mean we can entirely drop OCaml as a dependency, even in CI.

### PR DESCRIPTION
**Pull request by Timmmm**
This has the following benefits:

* Reduces the maintenance burden for making changes to the emulator. They don't have to be done in two places and you don't need to learn OCaml.
* Removes the dependency on OCaml and OPAM, which is a weirdly bug-prone tool.
* Removes confusion for new users about which emulator to use.
* Shorter CI time.

<strike>Note, I left the targets to actually build the OCaml code since I believe it is used by TestRig for its random instruction generation.</strike>

Fixes #509

### CODE REVIEW

**Comment by Alasdair** [comment]
The Sail to OCaml translation could also use quite a few changes, as it's the first thing I ever implemented when I started working on Sail. It still uses ocamlbuild for example, whereas the OCaml ecosystem has now mostly move to dune.

Removing the use here would actually free me to make some much needed changes.

**Comment by Timmmm** [comment]
@PeterRugg This MR removes OCaml support from the model, which I think would affect TestRig. Is that a deal breaker? What's the status of TestRig?

**Comment by PeterRugg** [comment]
Yeah, we certainly don't compare against the ocaml model, and can't see any reason why we would. Losing the sail-based test generation is sad, but as @jrtc27 says, we never really got chance to investigate using it properly.

**Review by arichardson — COMMENTED** [review]
I believe the one thing we are losing here is the ability to generate the DTB for the OS boot. I wonder if this should be added to the C simulator first? Otherwise I am more than happy to see the ocaml simulator being removed since it makes it a lot easier to add new extensions with hooks.

**Comment by arichardson** [comment]
> I believe the one thing we are losing here is the ability to generate the DTB for the OS boot. I wonder if this should be added to the C simulator first? Otherwise I am more than happy to see the ocaml simulator being removed since it makes it a lot easier to add new extensions with hooks.

Looks like it's just a simple string template (`let spike_dts isa_spec mmu_spec cpu_hz insns_per_rtc_tick mems`), so should be easy to move over to C.

**Comment by Timmmm** [comment]
I have actually implemented that for our C++ emulator so I'll port it over.

**Comment by Timmmm** [comment]
> I have actually implemented that for our C++ emulator so I'll port it over.

I actually haven't implemented it - I was thinking of something else. It *is* pretty simple; just generate [a string template](https://github.com/riscv/sail-riscv/blob/c61351e16e803d5ae8c14781d13a4bf214eb51d7/ocaml_emulator/platform_impl.ml#L69) and then [run](https://github.com/riscv/sail-riscv/blob/c61351e16e803d5ae8c14781d13a4bf214eb51d7/ocaml_emulator/platform_impl.ml#L166) `dtc` to compile it.

But doing that in C is a right pain, so I'm voting we defer this and add it again if/when somebody asks for it.

If you want to boot Linux you can still manually make a DTB.

**Comment by arichardson** [comment]
> > I have actually implemented that for our C++ emulator so I'll port it over.
> 
> I actually haven't implemented it - I was thinking of something else. It _is_ pretty simple; just generate [a string template](https://github.com/riscv/sail-riscv/blob/c61351e16e803d5ae8c14781d13a4bf214eb51d7/ocaml_emulator/platform_impl.ml#L69) and then [run](https://github.com/riscv/sail-riscv/blob/c61351e16e803d5ae8c14781d13a4bf214eb51d7/ocaml_emulator/platform_impl.ml#L166) `dtc` to compile it.
> 
> But doing that in C is a right pain, so I'm voting we defer this and add it again if/when somebody asks for it.
> 
> If you want to boot Linux you can still manually make a DTB.

Seems fine to me. It would be nice to generate this so that operating systems can parse it and boot, but can be done as a follow-up when needed.

**Review by arichardson — APPROVED** [review]

**Review by jordancarlin — APPROVED** [review]
LGTM

File: README.md [hunk 1]
```diff
@@ -36,7 +36,7 @@ What is Sail?
 engineer-friendly language, much like earlier vendor pseudocode, but more precisely defined and with tooling to support a wide range of use-cases.
 <p>
 
-Given a Sail specification, the tool can type-check it, generate documentation snippets (in LaTeX or AsciiDoc), generate executable emulators (in C or OCaml), show specification coverage, generate versions of the ISA for relaxed memory model tools, support automated instruction-sequence test generation, generate  theorem-prover definitions for
+Given a Sail specification, the tool can type-check it, generate documentation snippets (in LaTeX or AsciiDoc), generate an executable emulators, show specification coverage, generate versions of the ISA for relaxed memory model tools, support automated instruction-sequence test generation, generate  theorem-prover definitions for
```
Comment (arichardson) [inline README.md:39]:
```suggestion
Given a Sail specification, the tool can type-check it, generate documentation snippets (in LaTeX or AsciiDoc), generate executable emulators, show specification coverage, generate versions of the ISA for relaxed memory model tools, support automated instruction-sequence test generation, generate  theorem-prover definitions for
```
We still have rv32 and rv64 emulators so plural seems fine?

File: README.md [hunk 2]
```diff
@@ -139,13 +139,9 @@ mapping clause assembly = SRET() <-> "sret"
...
 
-The model builds OCaml and C emulators that can execute RISC-V ELF
+The model builds a C emulator that can execute RISC-V ELF
 files, and both emulators provide platform support sufficient to boot
-Linux, FreeBSD and seL4.  The OCaml emulator can generate its own
-platform device-tree description, while the C emulator currently
-requires a consistent description to be manually provided.  The C
-emulator can be linked against the Spike emulator for execution with
-per-instruction tandem-verification.
+Linux, FreeBSD and seL4. The C emulator can be linked against the Spike emulator for execution with per-instruction tandem-verification.
```
Comment (arichardson) [inline README.md:144]:
```suggestion
Linux, FreeBSD and seL4. The C emulator can be linked against the Spike emulator for execution with per-instruction tandem-verification.
```
Not for this PR, but is the Spike tandem verification actually still functional? If not it should probably be removed.

**Review by arichardson — APPROVED** [review]

**Comment by Timmmm** [comment]
I'll merge this in a few days if nobody objects.

**Pull request merged**